File: C_C++/Lecture_notes/Lecture-3/exp.c
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
typedef struct exp Exp;

struct exp
{
    enum tag
    {
        Var = 1,
        Neg = 2,
        Divide = 3
    } t;
    union
    {
        struct exp *arg1;
        char *var_name;
    };
    struct exp *arg2;
};

bool isEqual(Exp *a, Exp *b)
{
    if (a == b)
        return true;
    if (a->t != b->t)
        return false;
    if (a->t == 1 && (strcmp(a->var_name, b->var_name) == 0))
        return true;
    if (a->t == 2 && isEqual(a->arg1, b->arg1))
        return true;
    if (a->t == 3 && isEqual(a->arg1, b->arg1) && isEqual(a->arg2, b->arg2))
        return true;
    return false;
}

int main()
{
    Exp *a = (Exp *)malloc(sizeof(Exp));
    Exp *b = malloc(sizeof(Exp));
    a->t = 2;
    b->t = 2;
    a->arg1 = b;
    b->arg1 = b;
    printf("%d", isEqual(a, a));
}

File: CPL/JavaCodes/Exp.java
package CPL.JavaCodes;

public class Exp {
    private int tag;

    Exp(int tag) {
        this.tag = tag;
    }
}

class Var extends Exp {
    private char var_name;

    public Var(char var_name) {
        super(1);
        this.var_name = var_name;
    }
}

class Neg extends Exp {
    private Exp arg;

    public Neg(Exp arg) {
        super(2);
        this.arg = arg;
    }
}

class Divide extends Exp {
    private Exp arg1;
    private Exp arg2;

    public Divide(Exp arg1, Exp arg2) {
        super(3);
        this.arg1 = arg1;
        this.arg2 = arg2;
    }
}